### PR TITLE
nixos/mysql: add authentication option

### PR DESCRIPTION
**Pull request by Scriptkiddi**
Add an authentication option to be able to create mysql users that do
not authenticate using the unix_socket plugin, since this requires the
username to exist on the host system.
(https://mariadb.com/kb/en/library/authentication-plugin-unix-socket/)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by aanderse** [comment]
I was under the impression that `mysql` hashes are [not safe](https://security.stackexchange.com/questions/151998/the-security-of-the-password-function-in-mysql).

**Comment by Scriptkiddi** [comment]
Hm good point how should we deal with that? I could move the hashes to a file or add a warning. I want this option because I want to be able to add users that do not require a unix account on the host

**Comment by aanderse** [comment]
@Scriptkiddi I'm planning some work on the `mysql` and `postgresql` modules for `20.03`, partially motivated by [my own needs](https://github.com/NixOS/nixpkgs/issues/50403) and partially motivated by those of @talyz in [#66274](https://github.com/NixOS/nixpkgs/pull/66274#discussion_r319751420). I was jotting down a few ideas this morning after reading through your PR so how about we let this PR sit on the back burner a few days and I'll get back to you?

**Comment by aanderse** [comment]
@Scriptkiddi @florianjacob @talyz @ryantm *very* rough first pass at an addition to the `mysql` interface which could be mirrored with `postgresql`: https://github.com/NixOS/nixpkgs/compare/master...aanderse:sql?expand=1

Example usage:
```
services.mysql.ensureDatabases = [
  "db_a"
  "db_b"
  { name = "db_c"; }
  { name = "db_d"; owner = "timmy@localhost"; }
  { name = "db_e"; owner = "'aaron'@'%'"; }
];
services.mysql.ensureUsers = [
  # creates timmy@localhost who requires password to authenticate
  { name = "timmy";
    passwordFile = pkgs.writeText "password" "thisismypassword";
    ensurePermissions = { ... usual stuff here ... };
  }

  # creates a user who has access to all databases and WITH GRANT OPTION
  { name = "ryantm";
    isSuperUser = true;
  }

  # remote user
  { name = "aaron";
    host = "%";
    hashedPasswordFile = pkgs.writeText "hashed-password" "*ABCDEF HASHED PASSWORD";
    ensurePermissions = { ... usual stuff here ... };
  }

  # demonstrating backwards compatibility
  { name = "moodle";
    ensurePermissions = { ... usual stuff here ... };
  }
];
```

Note this is entirely backwards compatible with existing `mysql` interface.

**Comment by Infinisil** [comment]
`types.coercedTo` should work.

**Comment by aanderse** [comment]
> `types.coercedTo` should work.

Thanks @Infinisil, that looks better.

I'm still hoping to hear more opinions about whether this sort of change is something we actually **want** in nixos. Clearly @Scriptkiddi wants this change, as they opened this PR. I think I'm in favor of this change. Is anyone against these types of changes to nixos?

This might be a good time to ping @danbst, @aszlig, and @thoughtpolice as the idea is that these changes would make their way into the `postgresql` module as well. What do you think about this type of functionality in nixos? See [example](https://github.com/NixOS/nixpkgs/pull/68353#issuecomment-530636986) for usage, but basically deeper database and user provisioning than already exists in nixos.

**Review by grahamc — COMMENTED** [review]
As a rule I'm against options which add "best effort", imperative options to Nix. Deleting a  user from these settings doesn't delete the user from the system, and I think that is something we want to avoid in NixOS.

A good hint that the option either doesn't belong or doesn't function properly is having the word `ensure` in the name.

**Comment by aanderse** [comment]
@grahamc I understand your points and agree. While I think this PR provides value, maybe the value comes at too high a cost, and doesn't fit quite right into nixos?

I know @talyz was also interested in this functionality, so I'll ask @talyz and @Scriptkiddi if they have anything to add at this point?

**Comment by talyz** [comment]
@grahamc @aanderse I agree that it's definitely not ideal, but find it a necessary evil. The alternatives I can think of seem worse to me:
- all modules are responsible for provisioning their own database
- manual intervention needed
- dropping the modules requiring a database

Am I missing a better alternative?

Also, note that this is not actually about `ensureDatabases` or `ensureUsers`, since they already exist - it's about extending their functionality. The current alternative is to implement the necessary features locally in the module that needs it, like we did for [GitLab](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/gitlab.nix#L607-L624).

**Comment by petabyteboy** [comment]
I am joining the discussion, because I needed to extend the postgresql module to run a script after the creation of a database specified in ensureDatabases. I implemented this option in https://github.com/petabyteboy/nixpkgs/commit/d86208eb4ee23ca1f345c96b04bb0f7ce3a5d0bf and was made aware of this discussion by @aanderse 

* manual intervention needed
> Fine for me, this was my initial approach, but others did request that the database is created automatically and the schema is loaded into the database after initial creation, which required this change

* dropping the modules requiring a database
> ...

* all modules are responsible for provisioning their own database
> I definitely prefer less duplication of code in the modules

=> Maybe we can make it an internal option so that it is not directly exposed to users?

**Comment by alyssais** [comment]
I think manual intervention is probably okay, as long as its well documented. It does feel like it would detract somewhat from NixOS’s value proposition of “write all your configuration in configuration.nix” if a bunch of commands are needed to set up initial state as well, though.

**Comment by alyssais** [comment]
Alyssa Ross <notifications@github.com> writes:

> I think manual intervention is probably okay, as long as its well
> documented. It does feel like it would detract somewhat from NixOS’s
> value proposition of “write all your configuration in
> configuration.nix” if a bunch of commands are needed to set up initial
> state as well, though.

To phrase this another way, what's the difference between a package
creating its initial state on startup itself, and us doing it?  What
makes one okay and not the other?

**Comment by Scriptkiddi** [comment]
> As a rule I'm against options which add "best effort", imperative options to Nix. Deleting a user from these settings doesn't delete the user from the system, and I think that is something we want to avoid in NixOS.
> 
> A good hint that the option either doesn't belong or doesn't function properly is having the word `ensure` in the name.

Since I'm new to NixOs, couldnt we just modify the ensure script to delete users?

**Comment by flokli** [comment]
@Scriptkiddi we can't distinguish users that were configured via NixOS previously from users created by an admin, and don't want to delete these by accident, I assume.

**Comment by Scriptkiddi** [comment]
You could add a not delete list or something like it to, protect users that are added another way

**Comment by talyz** [comment]
> To phrase this another way, what's the difference between a package
creating its initial state on startup itself, and us doing it?  What
makes one okay and not the other?

Exactly. My solution to this would be to reimplement the missing functionality in my configuration to not have to do it manually (since I'm deploying systems with NixOps) and I'm sure others would do the same. We'd end up with the same thing, but worse.

**Comment by aanderse** [comment]
I found out this isn't the first PR to discuss in depth user provisioning for the `mysql` option. https://github.com/NixOS/nixpkgs/pull/6963

I'm still hoping @danbst and @aszlig will chime in with their opinions and reasoning on those opinions. @davidak, @copumpkin, and @edolstra may also have opinions on the topic of in depth user creation for the `mysql` module that they would be interested in sharing too.

**Comment by talyz** [comment]
I have a WIP for making `ensureDatabases` and `ensureUsers` declarative for the PostgreSQL module: #72365. Could you, and anyone else interested in the topic, have a look and see if this is something you'd like to see, @grahamc?

**Comment by florianjacob** [comment]
Regarding ensure* options: As the author of all `ensure*` options currently in master except the postgresql ones, I'm obviously thinking it's a good idea that helps users. I don't see why every module author and user should code the same functions again and again to create their databases, users, printers and whatever. Yes, those all are inherently stateful and probably never can be done fully declarative. But they just work, they provide you with automated printer setup in your fresh nixos install, they don't delete manually configured printers and require manual intervention on destructive operations, i.e. they will intentionally never delete your database, under any circumstance. That's why they just `ensure`, and do exactly that, and are helpful.

As this was one of my first big NixOS contributions, I chose the defensive prefix `ensure` for those kind of options, though many other options of this kind exist in NixOS, e.g. many service startup scripts that essentially only *ensure* the data directories are in a somewhat correct state and permissions, but obviously don't delete superfluous files or access permissions, or `config.users.users` would actually be `config.users.ensureUsers` if not set as mutable, which is the current default. Today I'm not sure anymore [whether the `ensure` prefixes really make sense](https://github.com/NixOS/nixpkgs/pull/55510) and whether the prefixes should be everywhere on such options as a warning, or nowhere.

If the community decides though that such options don't belong in NixOS and all users and module authors should go back and write their own bash scripts for that, I will accept that and stop proposung such options.

Regarding this specific PR: Seems like a good way to handle passwords for remote users (can't see why one should use that for anything that runs on the same host though) if a secure password hashing is used and / or the hash is stored in a non-accessible location. The default unsalted SHA1 hashing really isn't much better than plaintext passwords.
For more secure password hashing, state of the art seems to be to use [authentication plugins](https://mariadb.com/kb/en/library/authentication-plugins/) for both MariaDB and MySQL, bud sadly they don't have a common secure denominator.

MariaDB offers [https://mariadb.com/kb/en/authentication-plugin-ed25519/](authentication-plugin-ed25519) and MySQL offers [authentication-plugin-sha256](https://dev.mysql.com/doc/refman/8.0/en/sha256-pluggable-authentication.html) (why are both not using or at least offering [actual password hashing functions](https://en.wikipedia.org/wiki/Key_derivation_function)? No idea…)

File: nixos/modules/services/databases/mysql.nix [hunk 1]
```diff
@@ -153,6 +153,20 @@ in
                 Name of the user to ensure.
               '';
             };
+          authentication_option = mkOption {
+              type = types.str;
```
Comment (danbst) [inline nixos/modules/services/databases/mysql.nix]: I'd suggest a more specific type:
```
with types; enum ["socket" "password"]
```

File: nixos/modules/services/databases/mysql.nix [hunk 2]
```diff
@@ -406,7 +420,7 @@ in
 
                 ${concatMapStrings (user:
                   ''
-                    ( echo "CREATE USER IF NOT EXISTS '${user.name}'@'localhost' IDENTIFIED WITH ${if isMariaDB then "unix_socket" else "auth_socket"};"
+                    ( echo "CREATE USER IF NOT EXISTS '${user.name}'@'localhost' ${optionalString  (user.authentication_option == "password") ''IDENTIFIED BY PASSWORD '${user.password_hash}' ''} ${optionalString  (user.authentication_option == "socket") ''IDENTIFIED WITH  ${if isMariaDB then "unix_socket" else "auth_socket"} ''};"
```
Comment (danbst) [inline nixos/modules/services/databases/mysql.nix]: A few issues:
- line too long, should fit 100 columns
- what happens when password is changed? As I understand it, user already exists so his password will silently remain the same. This is bad. Please do run ALTER USER to change password hash when it was changed in NixOS config
- same for `authentication_option`

**Review by danbst — CHANGES_REQUESTED** [review]

**Comment by Scriptkiddi** [comment]
How are the chances of getting this merged? I'm willing to put in the work, but I don't want it to sit there open like the postgresql pr

**Comment by aanderse** [comment]
@Scriptkiddi sorry for the delay on this. We're scheduling a call to discuss the best way forward. More details to come.

**Comment by Scriptkiddi** [comment]
@aanderse Hey :), any news?

**Comment by aanderse** [comment]
@Scriptkiddi yes! @grahamc, @flokli, and myself had a chat about this last week. We all agreed that declarative, reproducible database provisioning in the NixOS module system is a good thing.

Currently we have declarative provisioning but we decided we want to take this a step further and make it reproducible, so that the `mysql` and `postgresql` modules can become true to the spirit of NixOS.

To that end we have decided a program/script needs to be written to handle user and database provisioning, in the same way that a program/script exists to manage system user accounts in NixOS. The program/script would take a `json` file (describing what database users and databases are desired) as input and then compare that against the existing database users and databases that `mysql` (or `postgresql`) already have and create the users required to achieve this result.

A couple notes...
- we want to add `services.mysql.mutableUsers` and `services.postgresql.mutableUsers` options which provides similar to functionality as `users.mutableUsers`
- while we want to delete database users which aren't declared in `configuration.nix` when `mutableUsers` is set to `false`, we never want to delete databases when they are not in `configuration.nix`, similar to how we never delete users home directories in the same scenario
- the program to handle user and database provisioning would ideally be written in `rust`
- we are looking for people to hack on this with us, especially the (ideally `rust`) program/script to create and delete database users, and create databases as described by `configuration.nix`
- it would be fantastic if we could have this ready for the `20.03` release... which is a tight timeline, but hey let's try our best right?

I'm really hoping a few people from this thread will be interested in hacking on this. @Scriptkiddi, @talyz, @florianjacob ... maybe @petabyteboy - any interest? `rust` skills in demand :smile: 

PS. @talyz since this is a fairly breaking change it is a great time to design module options from ground up! So all the goodness (`postgresql` extensions) that you were hoping for **will** happen with this change. I look forward to your contribution on this. Thanks for https://github.com/NixOS/nixpkgs/pull/75688 as a `bash` implementation.

**Comment by aanderse** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/94048.

**Pull request closed without merging**